### PR TITLE
server: support compressed MySQL protocol

### LINKED ISSUE #22605: JDBC useCompression=true does not work
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 4. What is your TiDB version? (Required)

Release Version: v4.0.0
Edition: Community
Git Commit Hash: 689a6b6439ae7835947fcaccf329a3fc303986cb
Git Branch: heads/refs/tags/v4.0.0
UTC Build Time: 2020-05-28 01:37:40
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number: close #22605

Problem Summary: Support deflate compression in the mysql protocol

This helps with:
- Performance: Especially data intensive work over low bandwidth connections
- Cost: This can reduce cost when bandwidth is payed for, e.g. in Cloud environments.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

```release-note
Support for the the compressed version of the MySQL protocol was added
```

### CODE REVIEW

**Comment by dveeden** [comment]
Tested with:

- `mysql --compress`
- `mysqlsh --compress`
- MySQL Connector/Python with `mysql.connector.connect(compress=True, ...)`

All the above with 8.x versions

**Comment by dveeden** [comment]
Tested with:

- MySQL Connector/J 8.0.30 and `useCompression=true`
- MySQL Client 8.0.30 and `mysql --compression-algorithms=zlib,zstd --zstd-compression-level=9 ...`
- MySQL Shell 8.0.29 and `mysqlsh --compression-algorithms=zstd --sql mysql://...`
- MySQL Connector/Python 8.0.21 and `mysql.connector.connect(compress=True, ...)`

**Comment by dveeden** [comment]
With my latest update it now compresses multiple mysql packets into a single compressed packet. 

This works somewhat:
```
$ mysql -h 127.0.0.1 -u root -P 4000 --ssl-mode=DISABLED --compression-algorithms=zlib 
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 405
Server version: 5.7.25-TiDB-v6.7.0-alpha-66-gc67761448-dirty TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible

Copyright (c) 2000, 2023, Oracle and/or its affiliates.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> \s
ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    407
Current database: *** NONE ***

--------------
mysql  Ver 8.0.32 for Linux on x86_64 (MySQL Community Server - GPL)

Connection id:		407
Current database:	
Current user:		root@127.0.0.1
SSL:			Not in use
Current pager:		stdout
Using outfile:		''
Using delimiter:	;
Server version:		5.7.25-TiDB-v6.7.0-alpha-66-gc67761448-dirty TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible
Protocol version:	10
Connection:		127.0.0.1 via TCP/IP
ERROR 2013 (HY000): Lost connection to MySQL server during query
No connection. Trying to reconnect...
Connection id:    409
Current database: *** NONE ***

Server characterset:	utf8mb4
Db     characterset:	utf8mb4
Client characterset:	utf8mb4
Conn.  characterset:	utf8mb4
TCP port:		4000
Protocol:		Compressed
Binary data as:		Hexadecimal
--------------

mysql> 
```

Next up is to figure out where we violate the protocol and fix that.

**Comment by dveeden** [comment]
This needs some tests, but otherwise it works.

Tested with:
- MySQL Client 8.0.33
- MySQL Shell 8.0.33
- MySQL Workbench 8.0.33
- MySQL Connector/Python 8.0.33 (with a fix for https://bugs.mysql.com/bug.php?id=110879)
- PHP 8.2.5 mysqlnd with `PDO::MYSQL_ATTR_COMPRESS`

This is with `--compression-algorithms=zstd`, `--compression-algorithms=zlib` and `--compression-algorithms=zstd,zlib` for MySQL Client and MySQL Shell. Other tools only seems to support zlib and not zstd.

The go-sql-driver/mysql doesn't support compression yet: https://github.com/go-sql-driver/mysql/issues/24

**Review by hawkingrei — APPROVED** [review]

File: server/packetio.go [hunk 1]
```diff
@@ -80,12 +93,48 @@ func (p *packetIO) setReadTimeout(timeout time.Duration) {
...
+			case mysql.CompressionZstd:
+				zstdReader, err := zstd.NewReader(p.bufReadConn, zstd.WithDecoderConcurrency(1))
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				zr = zstdReader.IOReadCloser()
+			default:
+				return nil, errors.New("Unknown compression algorithm")
+			}
+			r = zr
```
Comment (lance6716) [inline server/packetio.go:134]: can we directly assign to `r`?

File: server/packetio.go [hunk 2]
```diff
@@ -160,42 +213,164 @@ func (p *packetIO) writePacket(data []byte) error {
...
 		data[0] = 0xff
 		data[1] = 0xff
 		data[2] = 0xff
 
-		if n, err := p.bufWriter.Write(data[:4+mysql.MaxPayloadLen]); err != nil {
-			return errors.Trace(mysql.ErrBadConn)
-		} else if n != (4 + mysql.MaxPayloadLen) {
-			return errors.Trace(mysql.ErrBadConn)
+		if p.compressionAlgorithm != mysql.CompressionNone {
+			if n, err := p.compressedWriter.Write(data[:4+maxPayloadLen]); err != nil {
```
Comment (lance6716) [inline server/packetio.go:228]: seems we only use `Write` and `Flush` for both `compressedWriter` and `bufWriter`, maybe we can abstract an interface so there's no if-branch

File: server/packetio.go [hunk 3]
```diff
@@ -59,16 +62,26 @@ type packetIO struct {
 	// maxAllowedPacket is the maximum size of one packet in readPacket.
 	maxAllowedPacket uint64
 	// accumulatedLength count the length of totally received 'payload' in readPacket.
-	accumulatedLength uint64
+	accumulatedLength    uint64
+	compressionAlgorithm int
+	compressedSequence   uint8
+	zstdLevel            zstd.EncoderLevel
```
Comment (lance6716) [inline server/packetio.go:68]: will it be changed to other values than 3?

File: server/packetio.go [hunk 4]
```diff
@@ -160,42 +213,164 @@ func (p *packetIO) writePacket(data []byte) error {
...
+func (cw *compressedWriter) Write(data []byte) (int, error) {
+	return cw.buf.Write(data)
+}
+
+func (cw *compressedWriter) Flush() error {
+	var payload, compressedPacket bytes.Buffer
+	var w io.WriteCloser
+	var err error
+
+	minCompressLength := 50
```
Comment (lance6716) [inline server/packetio.go:312]: where did we get this value?

File: server/packetio.go (hunk 1 above)
Comment (dveeden) [reply server/packetio.go]: Yes: 6484a7efb1e7630d1c2ab364e0ff54da2fa788ae

File: server/packetio.go (hunk 3 above)
Comment (dveeden) [reply server/packetio.go:68]: Fixed in 16a3079114b4161b6e6db50c243bfadc51bcf51c

Also note that there is a translation here:

```go
	for i := 1; i <= 22; i++ {
		fmt.Printf("%d: %#v\n", i, zstd.EncoderLevelFromZstd(i))
	}
```
```
1: 1
2: 1
3: 2
4: 2
5: 2
6: 3
7: 3
8: 3
9: 3
10: 4
11: 4
12: 4
13: 4
14: 4
15: 4
16: 4
17: 4
18: 4
19: 4
20: 4
21: 4
22: 4
```

**Review by lance6716 — APPROVED** [review]

File: server/packetio.go (hunk 2 above)
Comment (lance6716) [reply server/packetio.go:227]: @dveeden I have no strong opinion about it. You can merge this PR or address it in following commit.

**Comment by xuyukeviki** [comment]
Can I implement this now？

**Pull request merged**